src/main/java/bio/terra/service/filedata/azure/tables/TableFileDao.java

```diff
@@ -0,0 +1,157 @@
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Component;
+
+/**
+ * TableFileDao provides CRUD operations on the "files" storage table in Azure. Objects in the file
+ * table are referred to by the dataset (owner of the files) and by any snapshots that reference the
+ * files. File naming is handled by the directory DAO. This DAO just handles the basic operations
+ * for managing the collection. It does not have logic for protecting against deleting files that
+ * have dependencies or creating files with duplicate paths.
+ */
+@Component
+public class TableFileDao {
+  private final Logger logger = LoggerFactory.getLogger(TableFileDao.class);
+  private final ExecutorService executor;
+  private static final String TABLE_NAME = "files";
+  private static final String PARTITION_KEY = "partitionKey";
+  private static final int SLEEP_BASE_SECONDS = 1;
+  private static final int AZURE_STORAGE_RETRIES = 1;
+
+  @Autowired
+  TableFileDao(@Qualifier("performanceThreadpool") ExecutorService executor) {
+    this.executor = executor;
+  }
+
+  public void createFileMetadata(TableServiceClient tableServiceClient, FireStoreFile newFile) {
+    tableServiceClient.createTableIfNotExists(TABLE_NAME);
+    TableClient tableClient = tableServiceClient.getTableClient(TABLE_NAME);
+    TableEntity entity = FireStoreFile.toTableEntity(PARTITION_KEY, newFile);
+    logger.info("creating file metadata for fileId: " + newFile.getFileId());
+    tableClient.createEntity(entity);
+  }
+
+  public boolean deleteFileMetadata(TableServiceClient tableServiceClient, String fileId) {
+    TableClient tableClient = tableServiceClient.getTableClient(TABLE_NAME);
+    TableEntity entity = tableClient.getEntity(PARTITION_KEY, fileId);
+    if (entity == null) {
+      return false;
+    }
+    logger.info("deleting file metadata for fileId " + fileId);
+    tableClient.deleteEntity(PARTITION_KEY, fileId);
+    return true;
+  }
+
+  public FireStoreFile retrieveFileMetadata(TableServiceClient tableServiceClient, String fileId) {
+    TableClient tableClient = tableServiceClient.getTableClient(TABLE_NAME);
+    TableEntity entity = tableClient.getEntity(PARTITION_KEY, fileId);
+    return FireStoreFile.fromTableEntity(entity);
+  }
+
+  /**
+   * Retrieve metadata from a list of directory entries.
+   *
+   * @param tableServiceClient An Azure table service client
+   * @param directoryEntries List of objects to retrieve metadata for
+   * @return A list of metadata object for the specified files. Note: the order of the list matches
+   *     with the order of the input list objects
+   */
+  List<FireStoreFile> batchRetrieveFileMetadata(
```

Comment:
Side not: I want to followon and try to ingest with a whole bunch of files to see what happens.  This seems like it'll make a ton of requests

Reply:
Not that I have a better idea :)